SpatialGDK/Source/SpatialGDK/Private/Interop/Startup/PartitionCreationStartupSteps.cpp

```diff
@@ -0,0 +1,318 @@
+
+		bQueryComplete = true;
+	}
+}
+
+bool FAuthCreateAndAssignPartitions::FDiscoverExistingPartitionsStep::TryFinish()
+{
+	PartitionQueryHandler.ProcessOps(Connection->GetWorkerMessages());
+	return bQueryComplete;
+}
+
+bool FAuthCreateAndAssignPartitions::FCreateNecessaryPartitionsStep::TryFinish()
+{
+	if (SharedState->DiscoveredPartitionEntityIds.Num() >= Setup.ExpectedServerWorkersCount)
+	{
+		return true;
+	}
+
+	const int32 PreexistingPartitionsCount = SharedState->DiscoveredPartitionEntityIds.Num();
+
+	const int32 PartitionsToSpawn = Setup.ExpectedServerWorkersCount - PreexistingPartitionsCount;
+	UE_LOG(LogSpatialStartupHandler, Log, TEXT("Spawning partition entities for %d virtual workers (need to spawn %d)"),
+		   Setup.ExpectedServerWorkersCount, PartitionsToSpawn);
+
+	TArray<Worker_EntityId> NewPartitionEntityIds;
+	NewPartitionEntityIds.Reserve(PartitionsToSpawn);
+	for (int32 PartitionIndex = 0; PartitionIndex < PartitionsToSpawn; ++PartitionIndex)
+	{
+		const VirtualWorkerId AssumedPartitionWorkerId = 1 + PreexistingPartitionsCount + PartitionIndex;
+		const Worker_EntityId PartitionEntityId = NetDriver->PackageMap->AllocateNewEntityId();
+		UE_LOG(LogSpatialStartupHandler, Log, TEXT("- Virtual Worker: %d. Entity: %lld. "), AssumedPartitionWorkerId, PartitionEntityId);
+		SpawnPartitionEntity(PartitionEntityId, AssumedPartitionWorkerId);
+		NewPartitionEntityIds.Emplace(PartitionEntityId);
+	}
+
+	SharedState->NewCreatedPartitionEntityIds = MoveTemp(NewPartitionEntityIds);
+	return true;
+}
+
+void FAuthCreateAndAssignPartitions::FCreateNecessaryPartitionsStep::SpawnPartitionEntity(Worker_EntityId PartitionEntityId,
+																						  VirtualWorkerId VirtualWorkerId)
+{
+	const bool bRunStrategyWorker = USpatialStatics::IsStrategyWorkerEnabled();
+	UAbstractLBStrategy* LoadBalanceStrategy = NetDriver->LoadBalanceStrategy;
+	const bool bDirectAssignment = bRunStrategyWorker && !LoadBalanceStrategy->IsStrategyWorkerAware();
+
+	UnrealServerInterestFactory* InterestF = nullptr;
+	if (!bRunStrategyWorker || bDirectAssignment)
+	{
+		InterestF = NetDriver->InterestFactory.Get();
+	}
+
+	QueryConstraint LBConstraint = LoadBalanceStrategy->GetWorkerInterestQueryConstraint(VirtualWorkerId);
+
+	TArray<FWorkerComponentData> Components = EntityFactory::CreatePartitionEntityComponents(
+		TEXT("WorkerPartition"), PartitionEntityId, InterestF, LBConstraint, VirtualWorkerId, NetDriver->DebugCtx != nullptr);
+
+	TArray<ComponentData> PartitionComponentsToCreate;
+	Algo::Transform(Components, PartitionComponentsToCreate, [](const FWorkerComponentData& Component) {
+		return ComponentData(OwningComponentDataPtr(Component.schema_type), Component.component_id);
+	});
+
+	Connection->SendCreateEntityRequest(MoveTemp(PartitionComponentsToCreate), PartitionEntityId, RETRY_UNTIL_COMPLETE);
+}
+
+bool FAuthCreateAndAssignPartitions::FWaitForPartitionVisibilityStep::TryFinish()
```

Comment:
Where is the polling done? Who updates the arrays that determine if we can stop?